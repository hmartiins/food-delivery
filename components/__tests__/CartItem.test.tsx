import { fireEvent, render } from '@testing-library/react-native';

import { CartItemType } from '@/type';

import { CartItem } from '../CartItem';

const mockIncreaseQty = jest.fn();
const mockDecreaseQty = jest.fn();
const mockRemoveItem = jest.fn();

jest.mock('@/stores/cart.store', () => ({
  useCartStore: () => ({
    increaseQty: mockIncreaseQty,
    decreaseQty: mockDecreaseQty,
    removeItem: mockRemoveItem,
  }),
}));

jest.mock('@/constants', () => ({
  images: {
    minus: 'minus-icon',
    plus: 'plus-icon',
    trash: 'trash-icon',
  },
}));

describe('<CartItem />', () => {
  const mockItem: CartItemType = {
    id: 'test-item-1',
    name: 'Burger Deluxe',
    price: 15.99,
    image_url: 'https://example.com/burger.jpg',
    quantity: 2,
    customizations: [
      {
        id: 'custom-1',
        name: 'Extra Cheese',
        price: 1.5,
        type: 'topping',
      },
    ],
  };

  beforeEach(() => jest.clearAllMocks());

  test('renders correctly the item information', () => {
    const { getByText } = render(<CartItem item={mockItem} />);

    expect(getByText('Burger Deluxe')).toBeTruthy();
    expect(getByText('$15.99')).toBeTruthy();
    expect(getByText('2')).toBeTruthy();
  });

  test('renders the product image correctly', () => {
    const { root } = render(<CartItem item={mockItem} />);
    const images = root.findAllByType('Image');
    const productImage = images[0];

    expect(productImage.props.source).toEqual({
      uri: 'https://example.com/burger.jpg',
    });
  });

  test('works correctly with item without customizations', () => {
    const itemWithoutCustomizations: CartItemType = {
      id: 'simple-item',
      name: 'Simple Burger',
      price: 12.99,
      image_url: 'https://example.com/simple-burger.jpg',
      quantity: 1,
    };

    const { getByText } = render(<CartItem item={itemWithoutCustomizations} />);

    expect(getByText('Simple Burger')).toBeTruthy();
    expect(getByText('$12.99')).toBeTruthy();
    expect(getByText('1')).toBeTruthy();
  });

  test('renders all icons correctly', () => {
    const { root } = render(<CartItem item={mockItem} />);
    const images = root.findAllByType('Image');

    expect(images[0].props.source).toEqual({
      uri: 'https://example.com/burger.jpg',
    });

    expect(images[1].props.source).toBe('minus-icon');
    expect(images[2].props.source).toBe('plus-icon');
    expect(images[3].props.source).toBe('trash-icon');
  });

  test('snapshot of the component', () => {
    const { toJSON } = render(<CartItem item={mockItem} />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('calls increaseQty when plus button is pressed', () => {
    const { getByTestId } = render(<CartItem item={mockItem} />);
    const plusButton = getByTestId('increase-qty-button');

    fireEvent.press(plusButton);

    expect(mockIncreaseQty).toHaveBeenCalledWith(
      mockItem.id,
      mockItem.customizations
    );
    expect(mockIncreaseQty).toHaveBeenCalledTimes(1);
  });

  test('calls decreaseQty when minus button is pressed', () => {
    const { getByTestId } = render(<CartItem item={mockItem} />);
    const minusButton = getByTestId('decrease-qty-button');

    fireEvent.press(minusButton);

    expect(mockDecreaseQty).toHaveBeenCalledWith(
      mockItem.id,
      mockItem.customizations
    );
    expect(mockDecreaseQty).toHaveBeenCalledTimes(1);
  });

  test('calls removeItem when trash button is pressed', () => {
    const { getByTestId } = render(<CartItem item={mockItem} />);
    const trashButton = getByTestId('remove-item-button');

    fireEvent.press(trashButton);

    expect(mockRemoveItem).toHaveBeenCalledWith(
      mockItem.id,
      mockItem.customizations
    );
    expect(mockRemoveItem).toHaveBeenCalledTimes(1);
  });
});
