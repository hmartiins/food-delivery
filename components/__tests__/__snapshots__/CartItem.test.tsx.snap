// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> snapshot of the component 1`] = `
<View
  className="cart-item"
>
  <View
    className="flex flex-row items-center gap-x-3"
  >
    <View
      className="cart-item__image"
    >
      <Image
        className="size-4/5 rounded-lg"
        resizeMode="cover"
        source={
          {
            "uri": "https://example.com/burger.jpg",
          }
        }
      />
    </View>
    <View>
      <Text
        className="base-bold text-dark-100"
      >
        Burger Deluxe
      </Text>
      <Text
        className="paragraph-bold mt-1 text-primary"
      >
        $
        15.99
      </Text>
      <View
        className="mt-2 flex flex-row items-center gap-x-4"
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="decrease-qty-button"
        >
          <Image
            className="size-1/2"
            resizeMode="contain"
            source="minus-icon"
            tintColor="#FF9C01"
          />
        </View>
        <Text
          className="base-bold text-dark-100"
        >
          2
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="increase-qty-button"
        >
          <Image
            className="size-1/2"
            resizeMode="contain"
            source="plus-icon"
            tintColor="#FF9C01"
          />
        </View>
      </View>
    </View>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="remove-item-button"
  >
    <Image
      className="size-5"
      resizeMode="contain"
      source="trash-icon"
    />
  </View>
</View>
`;
